query/plugins/org.eclipse.viatra.query.patternlanguage.emf/src/org/eclipse/viatra/query/patternlanguage/emf/types/EMFTypeInferrer.java

```diff
@@ -109,7 +110,9 @@ public Set<IInputKey> getAllPossibleTypes(Expression var) {
     }
 
     private TypeInformation collectConstraints(final Pattern pattern) {
-        final TypeInformation types = cache.get(this, pattern.eResource(), () -> new TypeInformation(typeSystem));
+        Resource eResource = pattern.eResource();
+        if (null == eResource) return new TypeInformation(typeSystem);
```

Comment:
I support your suggestion in principle - but then it should be tested thoroughly, keeping in mind that this makes the second half of the method reachable in a state where the pattern is detached from a Resource, which was not reachable and hence not tested so far.

Reply:
My main point here that without going to the second part of the method, type errors won't be found cross-patterns if the pattern parameters are untyped. Furthermore, patterns not in resources will not work with check/eval expressions at all (at least as far as I understand how Xbase works), so support for those cases should not be hidden in a completely unrelated PR.

Reply:
Yes, but patterns outside resources do not work if I understand correctly, and will remain nonworking after this change as well (whether my version or yours). The goal here is to make them not throw an NPE, not to make them functional.

Perhaps we should place an INFO or WARNING marker on Patterns at this point if we find them outside of a Resource?

Reply:
Those are also represented as EMF resources when loaded, so this case should not happen. The only case where it can happen that a Pattern is not in a Resource is if someone creates it using the EMF API directly.

Reply:
OK, I admit that I do not entirely know the conditions to provoke these, but I have definitely observed the NPE with the original code, without having constructed .

Reply:
As discussed, this was reverted from this PR; additionally, I have opened #189 to track the handling of this issue.